### ISSUE TITLE
Expose already registered settings to GraphQL

### ISSUE THREAD

**Issue by martin-braun** [wp-graphql/wp-graphql#1671]
I get a "Failed! token_revoked" when I try to join your Slack, so I want to ask here, sorry. I need to expose some custom settings of a foreign plugin in my WordPress page to GraphQL. I'm aware that I can [register custom settings on my own](https://www.wpgraphql.com/docs/settings/#registering-custom-settings), the problem is just that these are not my settings, but settings of a plugin that uses the [Redux Framework](https://github.com/reduxframework/redux-framework) to handle settings. 

I spotted the [`register_setting` call in the framework](https://github.com/reduxframework/redux-framework/blob/master/ReduxCore/framework.php#L2003). I debugged it, it will be reached once in the plugin and `$this->args['opt_name']` contains the `option_name` of the setting I want to expose to GraphQL, so it looks like I got the right spot. But, what's next?

For test purposes I changed the code to: 

```php
register_setting( $this->args['opt_name'] . '_group', $this->args['opt_name'], array(
    'type' => 'string',
    'sanitize_callback' => 'sanitize_text_field',
    'default' => NULL,
    'show_in_graphql' => true, // This tells WPGraphQL to show this setting in the Schema
    $this,
    '_validate_options'
) );
```

However, the setting does not get exposed to the GraphQL API. What can I do to make it work?

Also, once it works, how would I be able to modify it without opting out of updates? Unfortunately, Redux does not use actions on that part, so I could simply replace the `register_setting` call. Is there any suggested way to modify already registered settings to expose them to the GraphQL API?

Thank you for this wonderful plugin. Any help is highly appreciated.

**Comment by jasonbahl** [comment]
@martin-braun where are you registering your settings? if you register them within context such as `is_admin()` then WPGraphQL won't know about them. 

You'll want to register the settings in a more global context so that APIs that don't have context such as `is_admin()` can make use of the registry. 

You can `register_setting_fields()` within `is_admin()` context, because the fields themselves are only used within the admin context, but the setting registry and meta such as `sanitize_callback()` is useful to more contexts than just the admin.

**Comment by jasonbahl** [comment]
Ya, looks like Redux only registers settings in the context of the Admin: https://github.com/reduxframework/redux-framework/blob/master/ReduxCore/framework.php#L369

This means APIs that are not within the `is_admin()` scope cannot make use of those registered settings.

**Comment by martin-braun** [comment]
Big thanks for the quick response, really appreciate it. You are absolutely right in your observation. I just wonder why my above code did not work, when I was logged in as Admin while getting the schema in the GraphiQL IDE. 

May I ask how you would approach this issue on my spot?

I think the hard way would be to remove that action and build my own, in which I register the setting, every time. If you know of any alternative way to get things done, quicker and easier, I'd welcome it immensely.

**Comment by jasonbahl** [comment]
>  I just wonder why my above code did not work, when I was logged in as Admin while getting the schema in the GraphiQL IDE

The Schema doesn't change regardless of whether you're logged in or not. The Schema is a representation of what's _possible_ to ask for, regardless of the state of the user (authenticated or not). 

So much like how Post Types and Taxonomies are registered "globally" (and not _just_ in the Admin), in order for settings to show in the GraphQL Schema, the settings should also be registered "globally". 

> May I ask how you would approach this issue on my spot?

I worked up a quick prototype of what could become an official plugin for WPGraphQL for Redux at some point: https://gist.github.com/jasonbahl/bb1fb4a8e468d8312458b806b03197d4

You could take that and run with it, mapping all the Redux Fields to GraphQL Fields. 

It's a similar concept to WPGraphQL for Advanced Custom Field (https://github.com/wp-graphql/wp-graphql-acf) and you could steal any ideas from that codebase that would be helpful. 

Basically, the idea here is that GraphQL needs to read the registered settings in Redux, iterate over them, and map them to GraphQL Fields and Types. 

Each field type in Redux (text, image, color, date, etc) need to map to a GraphQL Field Type. 

In my quick example, it has all fields registered to GraphQL as a String, but you'd want to update it so an Image Field could return a GraphQL MediaItem, etc. 

I believe @dovy has some interest in adding support for GraphQL to Redux, so perhaps ya'll could work together to make something happen?

**Comment by martin-braun** [comment]
Dear @jasonbahl, you are just awesome, do you know that? =)

With your prototype I was able to pull this off the way I want and I have just to say "Thank you so much!". This really means a lot to me. I want to share my progress and the changes I made to pull this off, eventually.

First of all, the prototype uses `return`s in the `foreach` when it should `continue` to skip to the next element, instead of aborting the Redux section. Also, since I cannot register the Redux stuff by myself, I would not use such additional graphql fields on the Redux fields, but simply store an object with arrays outside the function, instead. Using the `plugins_loaded` action would also prevent the script to execute, so I got rid of that.

The way your prototype code is laid out, it would create a type for each Redux section into the `RootQuery`. This feels wrong, because I want the entire settings of the theme that uses Redux to be in one type that is on the `RootQuery`. So in theory, another plugin / theme could use the Redux framework, too, and it would get it's own exclusive `RootQuery` sub type.

Here is the code I ended up with. It shows how I defined an option in the `footer_copyright` section of the `MasterStudy` theme's Redux settings: 

```php
<?php

$all_exposed_settings = [
	'stm_option' => [
		'footer_copyright' => [
			'footer_copyright_text' => 'String'
		]
	]
];

if ( isset( $all_exposed_settings ) && ! empty( $all_exposed_settings ) ) {

	add_action( 'redux/loaded', function( $redux ) use( $all_exposed_settings ) {
		
		if ( isset( $redux->sections ) && ! empty( $redux->sections ) ) {
	
			$opt_name = $redux->args['opt_name'];
	
			if ( isset( $all_exposed_settings[$opt_name] ) ) {
	
				$gql_settings = $all_exposed_settings[$opt_name];
	
				$gql_root_type_name = $redux->args['display_name'] . 'Settings';
				$gql_root_type_name = str_replace( '_', '', ucwords( $gql_root_type_name, '_' ) );
	
				// console_log('Redux Sections', $opt_name, $redux->sections);
	
				$gql_section_fields = [];
	
				foreach ( $redux->sections as $section_id => $section ) {
	
					if ( isset( $gql_settings[$section['id']] ) ) {
	
						$gql_group_settings = $gql_settings[$section['id']];
	
						if ( isset( $section['fields'] ) && ! empty( $section['fields'] ) && is_array( $section['fields'] ) ) {
	
							$gql_field_fields = [];
	
							foreach ( $section['fields'] as $setting_field ) {
	
								if ( isset( $gql_group_settings[$setting_field['id']] ) ) {
						
									$gql_field_name = $setting_field['id'];
									$gql_field_name = str_replace( '_', '', ucwords( $gql_field_name, '_' ) );
									$gql_field_type = $gql_group_settings[$setting_field['id']];
	
									$gql_field_fields[$gql_field_name] = [
										'type' => $gql_field_type,
										'description' => isset( $setting_field['desc'] ) ? $setting_field['desc'] : '',
										'resolve' => function() use ( $opt_name, $redux, $setting_field ) {
											return Redux::getOption( $opt_name, $setting_field['id'] );
										}
									];
	
								}
	
							}
	
							if ( ! empty( $gql_field_fields ) ) {
	
								$gql_sub_type_name = $section['id'];
								$gql_sub_type_name = str_replace( '_', '', ucwords( $gql_sub_type_name, '_' ) );
	
								register_graphql_object_type( $gql_sub_type_name, [
									'fields' => $gql_field_fields
								] );
	
								$gql_section_fields[$gql_sub_type_name] = [
									'type' => $gql_sub_type_name,
									'resolve' => function() use ( $section ) {
										return $section;
									}
								];
	
							}
							
						}
	
					}
	
				}
	
				if ( ! empty( $gql_section_fields ) ) {
	
					register_graphql_object_type( $gql_root_type_name, [
						'fields' => $gql_section_fields
					] );
				
					register_graphql_field( 'RootQuery', $gql_root_type_name, [
						'type' => $gql_root_type_name,
						'resolve' => function() use ( $redux ) {
							return $redux->sections;
						}
					] );
	
				}
	
			}
			
		}
	
	} );

}
```

And here is the beautiful result: 

<img width="1139" alt="Screen Shot 2021-01-09 at 05 41 49" src="https://user-images.githubusercontent.com/24259245/104083258-7ab38480-523d-11eb-84ab-466d93e7e1a8.png">

What should I say? It even works in the front-end when I'm not logged in, besides the fact that the settings are only registered in `admin_init`. I won't say I understand the full magic behind this, but I can clearly say that your snippet gave me a good portion of knowledge when it comes to working with WP GraphQL and it works, I still can't believe it.

Happy weekend! :)
